File: flutter_inappwebview_windows/windows/types/web_history_item.cpp
#include "web_history_item.h"

namespace flutter_inappwebview_plugin
{
  WebHistoryItem::WebHistoryItem(const std::optional<int64_t>& entryId, const std::optional<int64_t>& index, const std::optional<int64_t>& offset,
    const std::optional<std::string>& originalUrl, const std::optional<std::string>& title,
    const std::optional<std::string>& url)
    : entryId(entryId), index(index), offset(offset), originalUrl(originalUrl), title(title), url(url)
  {}

  WebHistoryItem::WebHistoryItem(const flutter::EncodableMap& map)
    : WebHistoryItem(get_optional_fl_map_value<int64_t>(map, "entryId"),
      get_optional_fl_map_value<int64_t>(map, "index"),
      get_optional_fl_map_value<int64_t>(map, "offset"),
      get_optional_fl_map_value<std::string>(map, "originalUrl"),
      get_optional_fl_map_value<std::string>(map, "title"),
      get_optional_fl_map_value<std::string>(map, "url"))
  {}

  flutter::EncodableMap WebHistoryItem::toEncodableMap() const
  {
    return flutter::EncodableMap{
      {"entryId", make_fl_value(entryId)},
      {"index", make_fl_value(index)},
      {"offset", make_fl_value(offset)},
      {"originalUrl", make_fl_value(originalUrl)},
      {"title", make_fl_value(title)},
      {"url", make_fl_value(url)}
    };
  }
}

File: flutter_inappwebview_platform_interface/lib/src/types/web_history_item.dart
import 'package:flutter_inappwebview_internal_annotations/flutter_inappwebview_internal_annotations.dart';

import '../web_uri.dart';
import 'web_history.dart';
import 'enum_method.dart';

part 'web_history_item.g.dart';

///A convenience class for accessing fields in an entry in the back/forward list of a `WebView`.
///Each [WebHistoryItem] is a snapshot of the requested history item.
@ExchangeableObject()
class WebHistoryItem_ {
  ///Original url of this history item.
  WebUri? originalUrl;

  ///Document title of this history item.
  String? title;

  ///Url of this history item.
  WebUri? url;

  ///0-based position index in the back-forward [WebHistory.list].
  int? index;

  ///Position offset respect to the currentIndex of the back-forward [WebHistory.list].
  int? offset;

  ///Unique id of the navigation history entry.
  @SupportedPlatforms(platforms: [WindowsPlatform()])
  int? entryId;

  WebHistoryItem_({
    this.originalUrl,
    this.title,
    this.url,
    this.index,
    this.offset,
    this.entryId,
  });
}
